File: ios/App/App/AetherMediaIntegrityPlugin.swift
import Capacitor
import Foundation

@objc(AetherMediaIntegrityPlugin)
public class AetherMediaIntegrityPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AetherMediaIntegrityPlugin"
    public let jsName = "AetherMediaIntegrity"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "inspect", returnType: CAPPluginReturnPromise)
    ]

    private let worker = DispatchQueue(
        label: "org.castaliainstitute.aethertak.media-integrity",
        qos: .userInitiated
    )

    @objc func inspect(_ call: CAPPluginCall) {
        guard
            let value = call.getString("uri"),
            let source = URL(string: value),
            source.isFileURL
        else {
            call.reject(
                "A local captured-media file URI is required.",
                "INVALID_MEDIA_URI"
            )
            return
        }

        let resolved = source.resolvingSymlinksInPath().standardizedFileURL
        let appRoot = URL(fileURLWithPath: NSHomeDirectory(), isDirectory: true)
            .resolvingSymlinksInPath()
            .standardizedFileURL
        guard
            resolved.path != appRoot.path,
            resolved.path.hasPrefix(appRoot.path + "/")
        else {
            call.reject(
                "Only app-private captured media can be inspected.",
                "MEDIA_OUTSIDE_SANDBOX"
            )
            return
        }

        worker.async {
            do {
                let result = try inspectMediaFile(at: resolved)
                try FileManager.default.setAttributes(
                    [
                        .protectionKey:
                            FileProtectionType
                            .completeUntilFirstUserAuthentication
                    ],
                    ofItemAtPath: resolved.path
                )
                var values = URLResourceValues()
                values.isExcludedFromBackup = true
                var protectedFile = resolved
                try protectedFile.setResourceValues(values)
                call.resolve([
                    "sha256": result.sha256,
                    "sizeBytes": result.sizeBytes
                ])
            } catch {
                call.reject(
                    error.localizedDescription,
                    "MEDIA_INSPECTION_FAILED",
                    error
                )
            }
        }
    }
}

File: android/app/src/main/java/org/castaliainstitute/aethertak/field/plugins/AetherMediaIntegrityPlugin.kt
package org.castaliainstitute.aethertak.field.plugins

import android.net.Uri
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import java.io.File
import java.io.FileInputStream
import java.util.concurrent.Executors
import org.castaliainstitute.aethertak.field.media.inspectMedia

@CapacitorPlugin(name = "AetherMediaIntegrity")
class AetherMediaIntegrityPlugin : Plugin() {
    private val worker = Executors.newSingleThreadExecutor()

    @PluginMethod
    fun inspect(call: PluginCall) {
        val value = call.getString("uri")
        if (value.isNullOrBlank()) {
            call.reject("A local captured-media file URI is required.", "INVALID_MEDIA_URI")
            return
        }
        val source = localAppFile(value)
        if (source == null) {
            call.reject(
                "Only app-private captured media can be inspected.",
                "MEDIA_OUTSIDE_SANDBOX",
            )
            return
        }

        worker.execute {
            try {
                val result = FileInputStream(source).use(::inspectMedia)
                call.resolve(JSObject().apply {
                    put("sha256", result.sha256)
                    put("sizeBytes", result.sizeBytes)
                })
            } catch (error: Exception) {
                call.reject(
                    error.message ?: "Captured media inspection failed.",
                    "MEDIA_INSPECTION_FAILED",
                    error,
                )
            }
        }
    }

    override fun handleOnDestroy() {
        worker.shutdownNow()
        super.handleOnDestroy()
    }

    private fun localAppFile(value: String): File? {
        val parsed = Uri.parse(value)
        val candidate = when (parsed.scheme?.lowercase()) {
            null, "" -> File(value)
            "file" -> parsed.path?.let(::File)
            else -> null
        } ?: return null
        val resolved = runCatching { candidate.canonicalFile }.getOrNull() ?: return null
        val roots = listOf(context.filesDir, context.cacheDir, context.noBackupFilesDir)
            .mapNotNull { runCatching { it.canonicalFile }.getOrNull() }
        return resolved.takeIf { file ->
            file.isFile && roots.any { root ->
                file.path != root.path && file.path.startsWith(root.path + File.separator)
            }
        }
    }
}
